### PR TITLE
Collection of data and notification to listeners NETWORK_LATENCY [10791]

### PR DESCRIPTION
**Pull request by MiguelCompany**

### CODE REVIEW

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 1]
```diff
@@ -1310,6 +1311,62 @@ bool MessageReceiver::proc_Submsg_HeartbeatFrag(
...
+            if ((statistics_submessage_data_length != header.submessageLength) ||
+                    ((msg->pos + header.submessageLength) > msg->length))
+            {
+                break;
+            }
+
+            StatisticsSubmessageData data;
+            read_statistics_submessage(msg, data);
+            auto stats_participant = static_cast<StatisticsParticipantImpl*>(participant_);
+            stats_participant->on_network_statistics(source_guid_prefix_, source_locator, reception_locator, data);
```
Comment (MiguelBarro) [inline src/cpp/rtps/messages/MessageReceiver.cpp:1355]: The cast is unnecessary.

File: src/cpp/statistics/rtps/messages/RTPSStatisticsMessages.hpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
+// Copyright 2016 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (MiguelBarro) [inline src/cpp/statistics/rtps/messages/RTPSStatisticsMessages.hpp:1]: Update date

File: src/cpp/statistics/rtps/messages/RTPSStatisticsMessages.hpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    // Set current timestamp and sequence
+    auto submessage = (StatisticsSubmessageData*)(&send_buffer[statistics_pos]);
+    Time_t ts;
+    Time_t::now(ts);
+
+    submessage->ts.seconds = ts.seconds();
+    submessage->ts.fraction = ts.fraction();
+    submessage->seq.sequence = sequence.sequence;
+    submessage->seq.bytes = sequence.bytes;
+    submessage->seq.bytes_high = sequence.bytes_high;
```
Comment (MiguelBarro) [inline src/cpp/statistics/rtps/messages/RTPSStatisticsMessages.hpp:174]: submessage->seq = sequence;

Comment (MiguelBarro) [reply src/cpp/statistics/rtps/messages/RTPSStatisticsMessages.hpp:174]: Turns out the compiler optimization (msvc) copies the padding too. This padding is non-zeroed in stack objects and breaks RTPS standard directive.

File: src/cpp/statistics/rtps/StatisticsBase.cpp [hunk 4]
```diff
@@ -288,6 +288,37 @@ bool StatisticsParticipantImpl::remove_statistics_listener(
...
+        const rtps::StatisticsSubmessageData& data)
+{
+    static_cast<void>(source_participant);
+
+    using namespace eprosima::fastrtps::rtps;
+
+    Time_t ts(data.ts.seconds, data.ts.fraction);
+    Time_t current_ts;
+    Time_t::now(current_ts);
+    auto latency = static_cast<double>((current_ts - ts).to_ns());
```
Comment (MiguelBarro) [inline src/cpp/statistics/rtps/StatisticsBase.cpp:304]: use `float` to match the `Locator2LocatorData` setter.

**Review by MiguelBarro — CHANGES_REQUESTED** [review]
Please solve the minor issues

**Comment by MiguelCompany** [comment]
> Please solve the minor issues

Done in 673c41f2c3278cc37747d5008d17fa8ace9e1559

**Review by MiguelBarro — APPROVED** [review]
LGTM

**Pull request merged**